tests/lts_compatibility.py

```diff
@@ -0,0 +1,253 @@
+            args.oe_binary,
+            args.package,
+            library_dir=LOCAL_CHECKOUT_DIRECTORY,
+        )
+        network.consortium.add_new_code(primary, new_code_id)
+
+        # Add one more node than the current count so that at least one new
+        # node is required to reach consensus
+        # Note: alternate between joining from snapshot and replaying entire ledger
+        new_nodes = []
+        from_snapshot = True
+        for _ in range(0, len(network.get_joined_nodes()) + 1):
+            new_node = network.create_node(
+                "local://localhost",
+                binary_dir=LOCAL_CHECKOUT_DIRECTORY,
+                library_dir=LOCAL_CHECKOUT_DIRECTORY,
+                version=None,
+            )
+            network.join_node(new_node, args.package, args)
+            network.trust_node(new_node, args)
+            from_snapshot = not from_snapshot
+            new_nodes.append(new_node)
+
+        # Verify that all nodes run the expected CCF version
+        for node in network.get_joined_nodes():
+            if node not in lts_nodes or major_version > 1:
+                with node.client() as c:
+                    r = c.get("/node/version")
+                    assert r.body.json()["ccf_version"] == (
+                        args.ccf_version if node in new_nodes else lts_version
+                    )
+
+        issue_activity_on_live_service(network, args)
+
+        # Test that new nodes can become primary with LTS nodes as backups
+        # Note: Force a new node as primary by isolating LTS nodes
+        for node in lts_nodes:
+            node.suspend()
+
+        new_primary, _ = network.wait_for_new_primary(primary, nodes=new_nodes)
+        assert (
+            new_primary in new_nodes
+        ), "New node should have been elected as new primary"
+
+        for node in lts_nodes:
+            node.resume()
+
+        try:
+            network.wait_for_new_primary(new_primary)
+            assert False, "No new primary should be elected while"
+        except TimeoutError:
+            pass
+
+        # Retire one new node, so that at least one LTS node is required to reach consensus
+        other_new_nodes = [
+            node
+            for node in network.get_joined_nodes()
+            if (node is not new_primary and node in new_nodes)
+        ]
+        network.retire_node(new_primary, other_new_nodes[0])
+
+        issue_activity_on_live_service(network, args)
+
+        # Retire LTS nodes and code id
+        old_code_id = infra.utils.get_code_id(
+            args.enclave_type,
+            args.oe_binary,
+            args.package,
+            library_dir=lts_library_dir,
+        )
+        network.consortium.retire_code(primary, old_code_id)
+        for node in lts_nodes:
+            network.retire_node(new_primary, node)
+            node.stop()
+
+        issue_activity_on_live_service(network, args)
+
+
+def run_ledger_compatibility_since_first(args, use_snapshot):
```

Comment:
TODO: 
- [ ] Test JWTs.
- [ ] Since KV table schemas are only used when a table is read, a successful recovery doesn't mean that the new version can handle the old schema. So after each recovery, we should issue requests that read those tables (which may be tricky for some).

Reply:
Only for built-in functionality though, not for application tables (eg. logging sample).